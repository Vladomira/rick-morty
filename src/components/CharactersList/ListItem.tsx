import Image from 'next/image';
import Link from 'next/link';

import styles from '../../../styles/CharactersList.module.scss';

import { CharacterListItem } from '../../types/CharactersData';

export const ListItem = ({ id, name, image }: CharacterListItem) => {
  return (
    <li className={styles.characters__item}>
      <Link href={`/characters/${id}`} className={styles.characters__link}>
        <div className={styles.characters__thumb}>
          <Image
            className={styles.characters__img}
            src={image}
            width={210}
            height={210}
            alt={name}
          />
        </div>
        <p className={styles.characters__text}>{name}</p>
      </Link>
    </li>
  );
};
