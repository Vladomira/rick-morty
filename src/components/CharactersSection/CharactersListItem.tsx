import { CharacterListItem } from "@/src/types/CharactersData";
import Image from "next/image";
import Link from "next/link";

export const CharactersListItem = ({ id, name, image }: CharacterListItem) => {
  return (
    <li className="characters__item" role="list-item">
      <Link href={`/characters/${id}`} className="characters__link">
        <div className="characters__thumb">
          <Image
            className="characters__img"
            src={image}
            width={210}
            height={210}
            alt={name}
          />
        </div>
        <div className="characters__text-box">
          <p className="characters__text">{name}</p>
        </div>
      </Link>
    </li>
  );
};
