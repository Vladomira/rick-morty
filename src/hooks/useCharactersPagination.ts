import { useState } from "react";

import { OperationVariables, useSuspenseQuery } from "@apollo/client";

import { incrementPage } from "../helpers/incrementPage";
import { characters } from "../service/queries/queries.graphql";
import { CharacterListItem } from "../types/CharactersData";
import { Count } from "../types/domain";

type CharacterQueryResult = {
  characters: { info: Count; results: CharacterListItem[] };
};

export const useCharactersPagination = (
  initialData: CharacterListItem[],
  count: number
) => {
  const [page, setPage] = useState<number>(2);
  const [items, setItems] = useState<CharacterListItem[]>(initialData);
  const { data, error } = useSuspenseQuery<
    CharacterQueryResult,
    OperationVariables
  >(characters, {
    variables: { page },
    skip: page === 1,
  });

  const itemsPerPage = 20;
  const pagesCount = Math.floor(count / itemsPerPage);

  const getMoreCharacters = () => {
    incrementPage({ page, setPage, pagesCount });
    const results = data?.characters.results;

    results && setItems((prev) => [...prev, ...results]);
  };
  return {
    items,
    getMoreCharacters,
    hasMore: page < pagesCount,
    error,
  };
};
