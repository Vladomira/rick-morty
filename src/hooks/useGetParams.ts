import { useSearchParams } from "next/navigation";

export const useGetParams = (name: string) => {
  const searchParams = useSearchParams();
  return searchParams.get(name)?.toString();
};
