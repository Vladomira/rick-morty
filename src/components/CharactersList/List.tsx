import { ListItem } from './ListItem';
import styles from '../../../styles/CharactersList.module.scss';
import { CharacterFullData } from '../../types/CharactersData';

export const List = ({ items }: { items: CharacterFullData[] }) => {
  return (
    <ul className={styles.characters__list}>
      {items.map(({ id, name, image }) => {
        return <ListItem id={id} key={id} name={name} image={image} />;
      })}
    </ul>
  );
};
