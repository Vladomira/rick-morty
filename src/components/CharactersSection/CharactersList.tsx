import { CharactersListItem } from "./CharactersListItem";
import { CharacterListItem } from "@/src/types/CharactersData";

export const CharactersList = ({ items }: { items: CharacterListItem[] }) => {
  return (
    <ul className={"characters__list"} data-testid="characters-list">
      {items.map(({ id, name, image }) => (
        <CharactersListItem id={id} key={id} name={name} image={image} />
      ))}
    </ul>
  );
};
