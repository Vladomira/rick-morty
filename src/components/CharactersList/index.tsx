'use client';
import axios from 'axios';
import React, { useState } from 'react';
import InfiniteScroll from 'react-infinite-scroll-component';
import { BallTriangle } from 'react-loader-spinner';

import { List } from './List';
import style from '../../../styles/Home.module.scss';
import { baseURL } from '../../service/api';
import { CharacterFullData } from '../../types/CharactersData';
import { Condition } from '../../types/common';
type CharactersListProps = {
  characters: CharacterFullData[];
};

const CharactersList = ({ characters }: CharactersListProps) => {
  const [page, setPage] = useState<number>(2);
  const [items, setItems] = useState<CharacterFullData[]>(characters);
  const [loadMore, setLoadMore] = useState(true);
  const [condition, setCondition] = useState<Condition>('idle');

  const getMoreCharacters = async () => {
    setCondition('pending');
    if (page < 41) {
      setPage(prev => {
        return prev + 1;
      });
    } else {
      return;
    }
    const { data } = await axios.get(`${baseURL}/?page=${page}`);
    if (!data) {
      setCondition('rejected');
      return console.error('something went wrong');
    }
    setTimeout(() => {
      setItems((prev: CharacterFullData[]) => [...prev, ...data.results]);
    }, 2000);
    setCondition('resolved');
  };
  return (
    <InfiniteScroll
      dataLength={items.length}
      next={getMoreCharacters}
      scrollThreshold={'100%'}
      hasMore={loadMore}
      loader={
        <div className={style.preloader}>
          <BallTriangle color="yellow" height={100} width={90} />
        </div>
      }
      endMessage={<h4 className={style.header__title}>Nothing more to show</h4>}
    >
      <List items={items} />
    </InfiniteScroll>
  );
};

export default CharactersList;
