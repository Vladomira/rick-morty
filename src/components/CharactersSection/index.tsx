"use client";

import React, { useRef } from "react";

import { CharactersList } from "./CharactersList";
import { CharactersListProps } from "@/src/types/CharactersData";
import { AnimatePresence } from "framer-motion";
import InfiniteScroll from "react-infinite-scroll-component";
import { BallTriangle } from "react-loader-spinner";

import { useCharactersPagination } from "@/src/hooks/useCharactersPagination";

import InfoMessage from "../InfoMessage";
import ScrollButtons from "../ScrollButtons.tsx";

const CharactersSection = ({ charactersData, count }: CharactersListProps) => {
  const { items, getMoreCharacters, hasMore, error } = useCharactersPagination(
    charactersData,
    count
  );
  const sectionRef = useRef<HTMLDivElement | null>(null);

  return (
    <div className={"container m-auto"} ref={sectionRef}>
      {error && <InfoMessage message={error.message} />}
      {!error && items?.length > 0 && (
        <>
          <InfiniteScroll
            dataLength={items.length}
            next={getMoreCharacters}
            scrollThreshold={"95%"}
            hasMore={hasMore}
            loader={
              <div className="preloader__box">
                <BallTriangle color="yellow" height={100} width={90} />
              </div>
            }
            endMessage={<InfoMessage message="Nothing more to show" />}
          >
            <CharactersList items={items} />
          </InfiniteScroll>
          <AnimatePresence>
            {items.length > 20 && <ScrollButtons sectionRef={sectionRef} />}
          </AnimatePresence>
        </>
      )}
    </div>
  );
};

export default CharactersSection;
