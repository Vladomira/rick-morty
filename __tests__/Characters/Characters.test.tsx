import React from "react";

import {
  CHARACTERS_LIST_ID,
  LIST_ITEM_ROLE,
  mockCharactersData,
  mockErrorMessage,
} from "./constants";
import {
  characters,
  mockCharacters,
  mockReturnData,
  mockReturnDataError,
} from "./mock";
import { MockedProvider } from "@apollo/client/testing";
import {
  act,
  cleanup,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";

import { useCharactersPagination } from "@/src/hooks/useCharactersPagination";

import CharactersSection from "@/src/components/CharactersSection";
import { CharactersList } from "@/src/components/CharactersSection/CharactersList";
import ScrollButtons from "@/src/components/ScrollButtons.tsx";

jest.mock("@/src/hooks/useCharactersPagination");

const mockUseCharactersPagination =
  useCharactersPagination as jest.MockedFunction<
    typeof useCharactersPagination
  >;

mockUseCharactersPagination.mockReturnValue(mockReturnData);

describe("Characters", () => {
  afterEach(() => {
    cleanup();
  });

  // list
  it("should render a characters list and should have a length equal 2", () => {
    render(<CharactersList items={mockCharactersData} />);

    const list = screen.getByTestId(CHARACTERS_LIST_ID);

    expect(list).toBeInTheDocument();
    expect(list.children).toHaveLength(2);
    expect(list).toMatchSnapshot();
  });

  it("items should have the right value of  prop 'name'", () => {
    render(<CharactersList items={mockCharactersData} />);

    const items = screen.getAllByRole(LIST_ITEM_ROLE);
    const charactersNames = items.map((item) => item.textContent);

    expect(charactersNames).toEqual(["Rick Sanchez", "Morty Smith"]);
  });

  // component
  test("renders CharactersSection component", async () => {
    const { container } = render(
      <MockedProvider mocks={mockCharacters} addTypename={false}>
        <CharactersSection charactersData={[]} count={2} />
      </MockedProvider>
    );

    expect(await screen.findByText("Rick Sanchez")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("loads more characters on scroll", () => {
    render(
      <CharactersSection charactersData={mockReturnData.items} count={2} />
    );

    fireEvent.scroll(window, { target: { scrollY: 1000 } });

    expect(mockReturnData.getMoreCharacters).toHaveBeenCalled();
  });

  test("displays end message when no more characters", () => {
    mockUseCharactersPagination.mockReturnValue({
      ...mockReturnData,
      hasMore: false,
    });

    render(
      <CharactersSection charactersData={mockReturnData.items} count={2} />
    );

    expect(screen.getByText("Nothing more to show")).toBeInTheDocument();
  });

  test("displays error message on query error", async () => {
    mockUseCharactersPagination.mockReturnValue(mockReturnDataError);
    const errorMocks = [
      {
        request: {
          query: characters,
          variables: { page: 1 },
        },
        error: new Error(mockErrorMessage),
      },
    ];

    render(
      <MockedProvider mocks={errorMocks} addTypename={false}>
        <CharactersSection charactersData={[]} count={2} />
      </MockedProvider>
    );

    await waitFor(() => {
      expect(screen.getByRole("info-message")).toBeInTheDocument();
    });
  });

  it("scroll buttons should be in the document", async () => {
    const sectionRef = { current: document.createElement("div") };

    render(
      <div ref={sectionRef}>
        <ScrollButtons sectionRef={sectionRef} />
      </div>
    );

    act(() => {
      fireEvent.scroll(window, { target: { scrollY: 1000 } });
    });

    expect(screen.getByTestId("scroll-buttons")).toBeInTheDocument();

    await waitFor(
      () => {
        expect(screen.getByTestId("btn-top")).toBeInTheDocument();
        expect(screen.getByTestId("btn-bottom")).toBeInTheDocument();
      },
      { timeout: 1000 }
    );
  });
});
