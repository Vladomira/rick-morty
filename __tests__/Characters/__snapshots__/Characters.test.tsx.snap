// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters renders CharactersSection component 1`] = `
<div>
  <div
    class="container m-auto"
  >
    <div
      class="infinite-scroll-component__outerdiv"
    >
      <div
        class="infinite-scroll-component "
        style="height: auto; overflow: auto;"
      >
        <ul
          class="characters__list"
          data-testid="characters-list"
        >
          <li
            class="characters__item"
            role="list-item"
          >
            <a
              class="characters__link"
              href="/characters/1"
            >
              <div
                class="characters__thumb"
              >
                <img
                  alt="Rick Sanchez"
                  class="characters__img"
                  data-nimg="1"
                  decoding="async"
                  height="210"
                  loading="lazy"
                  src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=640&q=75"
                  srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=640&q=75 2x"
                  style="color: transparent;"
                  width="210"
                />
              </div>
              <div
                class="characters__text-box"
              >
                <p
                  class="characters__text"
                >
                  Rick Sanchez
                </p>
              </div>
            </a>
          </li>
          <li
            class="characters__item"
            role="list-item"
          >
            <a
              class="characters__link"
              href="/characters/2"
            >
              <div
                class="characters__thumb"
              >
                <img
                  alt="Morty Smith"
                  class="characters__img"
                  data-nimg="1"
                  decoding="async"
                  height="210"
                  loading="lazy"
                  src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=640&q=75"
                  srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=640&q=75 2x"
                  style="color: transparent;"
                  width="210"
                />
              </div>
              <div
                class="characters__text-box"
              >
                <p
                  class="characters__text"
                >
                  Morty Smith
                </p>
              </div>
            </a>
          </li>
        </ul>
        <div
          class="preloader__box"
        >
          <div
            aria-busy="true"
            aria-label="ball-triangle-loading"
            data-testid="ball-triangle-loading"
            role="status"
            style="display: flex;"
          >
            <svg
              data-testid="ball-triangle-svg"
              height="100"
              stroke="yellow"
              viewBox="0 0 57 57"
              width="90"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="none"
                fill-rule="evenodd"
              >
                <g
                  stroke-width="2"
                  transform="translate(1 1)"
                >
                  <circle
                    cx="5"
                    cy="50"
                    r="5"
                  >
                    <animate
                      attributeName="cy"
                      begin="0s"
                      calcMode="linear"
                      dur="2.2s"
                      repeatCount="indefinite"
                      values="50;5;50;50"
                    />
                    <animate
                      attributeName="cx"
                      begin="0s"
                      calcMode="linear"
                      dur="2.2s"
                      repeatCount="indefinite"
                      values="5;27;49;5"
                    />
                  </circle>
                  <circle
                    cx="27"
                    cy="5"
                    r="5"
                  >
                    <animate
                      attributeName="cy"
                      begin="0s"
                      calcMode="linear"
                      dur="2.2s"
                      from="5"
                      repeatCount="indefinite"
                      to="5"
                      values="5;50;50;5"
                    />
                    <animate
                      attributeName="cx"
                      begin="0s"
                      calcMode="linear"
                      dur="2.2s"
                      from="27"
                      repeatCount="indefinite"
                      to="27"
                      values="27;49;5;27"
                    />
                  </circle>
                  <circle
                    cx="49"
                    cy="50"
                    r="5"
                  >
                    <animate
                      attributeName="cy"
                      begin="0s"
                      calcMode="linear"
                      dur="2.2s"
                      repeatCount="indefinite"
                      values="50;50;5;50"
                    />
                    <animate
                      attributeName="cx"
                      begin="0s"
                      calcMode="linear"
                      dur="2.2s"
                      from="49"
                      repeatCount="indefinite"
                      to="49"
                      values="49;5;27;49"
                    />
                  </circle>
                </g>
              </g>
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Characters should render a characters list and should have a length equal 2 1`] = `
<ul
  class="characters__list"
  data-testid="characters-list"
>
  <li
    class="characters__item"
    role="list-item"
  >
    <a
      class="characters__link"
      href="/characters/1"
    >
      <div
        class="characters__thumb"
      >
        <img
          alt="Rick Sanchez"
          class="characters__img"
          data-nimg="1"
          decoding="async"
          height="210"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=640&q=75"
          srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=640&q=75 2x"
          style="color: transparent;"
          width="210"
        />
      </div>
      <div
        class="characters__text-box"
      >
        <p
          class="characters__text"
        >
          Rick Sanchez
        </p>
      </div>
    </a>
  </li>
  <li
    class="characters__item"
    role="list-item"
  >
    <a
      class="characters__link"
      href="/characters/2"
    >
      <div
        class="characters__thumb"
      >
        <img
          alt="Morty Smith"
          class="characters__img"
          data-nimg="1"
          decoding="async"
          height="210"
          loading="lazy"
          src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=640&q=75"
          srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F2.jpeg&w=640&q=75 2x"
          style="color: transparent;"
          width="210"
        />
      </div>
      <div
        class="characters__text-box"
      >
        <p
          class="characters__text"
        >
          Morty Smith
        </p>
      </div>
    </a>
  </li>
</ul>
`;
